README.md
# plotVR - Walk Through Your Data

This is a prototype to get your data into a Google Cardboard (VR) or into Augmented Reality (AR) and navigate using the computer keyboard.

> This package was presented at EuroPython 2019.
> Check out the recording video on <https://youtu.be/lxXC76jHc98?t=5h20m>. *The video covers the whole day in that room - the talk starts at 5h20m and ends at 5h46m.*

Are you bored by 3D-plots that only give you a simple rotatable 2d-projection? plotVR provides a simple way for data scientists to plot data, pick up a phone, get a real 3d impression - either by VR or by AR - and use the computer's keyboard to walk through the scatter plot (see [live demo](https://thomann.github.io/plotVR/plotVR/inst/)):

![Overview](images/overview.png?raw=true "Overview")

The technologies beneath this project are: a web server that handles the communication between the DataScience-session and the phone, WebSockets to quickly proxy the keyboard events, QR-codes facilitate the simple pairing of both, and an HTML-Page on the computer to grab the keyboard events. And the translation of these keyboard events into 3D terms is a nice exercise in three.js, OpenGL, and SceneKit for HTML, Android, and iOS resp.

> **Warning:** All data is transmitted unencrypted and everybody can connect! Please be carefule with private data!

> **Disclaimer:** This package is provided as-is and every usage is on your own ris.

## Installation

Install the R-package:
```r
devtools::install_github('thomann/plotVR',subdir='plotVR-R')
```
or the Python package:
```bash
pip install --upgrade "git+https://github.com/thomann/plotVR#egg=plotvr&subdirectory=plotVR-py"
```
> The installation from sources currently needs symbolic links to work, which is a bit of an issue on windows (see below). If you cannot make it work, wait to obtain precompiled wheels.

Also check out the native mobile apps:
- iOS: <https://github.com/thomann/PlotAR-ios>
- Android: <https://github.com/thomann/plotVR-android>


## Quick Tour

Working with the package is meant to be seamless - but it is a little hard to explain in text. Please check out the above [video](https://youtu.be/lxXC76jHc98?t=5h27m) at 5h27m where the demo starts.

Here we will describe how to plot the data in RStudio or Jupyter, then view it on your device, and finally navigate through the scene.

### Plot your data

In R load and plot the first three dimensions of `iris` with `iris$Species` as color:
```r
library(plotVR)
startExternalServerProcess()
plotVR(iris[,1:3],iris$Species)
```
In Python - or better even in Jupyter Lab - enter the following:
```python
import plotvr
plotvr.start_server_process()
from sklearn import datasets
iris = datasets.load_iris()
plotvr.plotvr(iris.data, iris.target)
```
This will look something like
![RStudio](images/screen-rstudio.png?raw=true)
and
<center><img src="images/screen-jupyter.png?raw=true" width=250></center>

### View in a browser or on a device

Now you can open the advertised webpage - also on your mobile device using any QR-code reader:
```
http://<ip-address of your machine>:2908/index.html
```
This will look something like (see [live demo](https://thomann.github.io/plotVR/plotVR/inst/)):

![VR view](images/screen-vr.png?raw=true "VR view")

> Naturally, you also can visit the server page on your computer: <http://localhost:2908/plotVR.html>.

Tap on the screen to bring it to full-screen!

### Keyboard Navigation

Now in order to navigate through your scene you can use the keyboard. In order to capture key events click on the display keyboard:
![Keyboard](images/screen-keyboard.png?raw=true)

They keyboard warns if it loses focus:
![Keyboard without focus](images/screen-keyboard-nofocus.png?raw=true)

This window also will get the keyboard focus when it opens. You now can use your computer keyboard to navigate through the virtual space (like in Quake!):

|  Key              |  Movement             |
|-------------------|-----------------------|
|   w               |  forward              |
|   s               | backward              |
|   a               |    left               |
|   d               |   right               |
|   q               |     up                |
|   e               |   down                |
|  up/down-arrow    |  rotate vertically    |
|  left/right-arrow |  rotate horizontally  |
|   r               |  reload               |
|  space            |  toggle velocity      |

### 

Now if you plot something new it should be reloaded automatically in your viewer:
```r
plotVR(trees)
```

## Issues

* This is not very stable at the moment!
* This repo uses symlinks which for Windows explicitely need to be activated, see <https://github.com/git-for-windows/git/wiki/Symbolic-Links>.
* There need to be better keyboard assignments, and maybe some more interface in the keyboard focus window.
* The default values are not well set, e.g. the starting point of the virtual observer.
* The file `plotVR-matlab/plotVR.m` can be used in MATLAB if you have access to a running server in Python or R.

## Acknowledgements

* For the WebVR-client: [`three.js`](http://threejs.org),
  [`polyfill`](https://github.com/googlevr/webvr-polyfill), and
  [`webvr-boilerplate`](https://github.com/borismus/webvr-boilerplate)
  (all needed parts included in this repository under `plotVR/inst/js/third-party`)
* [`httpuv`](https://github.com/rstudio/httpuv) for the websocket-server implementation in R
* [`tornado`](https://www.tornadoweb.org/) for the websocket-server implementation in Python
* <https://vr.google.com/cardboard/> for the cardboard!


plotVR-py/HISTORY.md
=======
History
=======

0.1.0 (2019-07-11)
------------------

* First Version


plotVR/R/caRdboard.R

# the state of the controller:
pkg.env <- new.env()

# list of all connected clients
pkg.env$clients <- list()

# the handle of the httpuv-server daemon
pkg.env$server_handle <- NULL

# the model to serve to devices
pkg.env$vr_data_json <- NULL

# interval for blocking server to serve at once
pkg.env$interruptIntervalMs <- 250

log.info <- function(...){
  try({
    if(getOption("plotVR.log.info", default=TRUE))
      cat(format(Sys.time(), "[%Y-%m-%d %H:%M:%S]"), "plotVR info:",...,fill=T)
  })
  try({
    filename <- getOption("plotVR.log.info.file")
    if(!is.null(filename))
      cat("plotVR info:",...,fill=T,file=filename)
  })
}

# ' @importFrom datasets iris
defaultDataJson <- function()
  plotVR(datasets::iris[,1:3],col=datasets::iris$Species, name="Iris", .send=FALSE, doOpenController=F)


process_request <- function(req, base="~/density/vr/") {
  log.info("Serving",req$PATH_INFO)
  # print(ls.str(req))
  log.info(req$REQUEST_METHOD)
  if(req$REQUEST_METHOD=="POST"){
    return(handle_upload_data(req))
  }
  path <- req$PATH_INFO
  wsUrl = paste0('ws://', ifelse(is.null(req$HTTP_HOST), req$SERVER_NAME, req$HTTP_HOST),'/ws')
  if(path=="/") path <- "/index.html"
  if(path=="/plotVR.html") path <- "/index.html"
  if( path == '/echo' )
    return(handle_echo(wsUrl))
  if(path == "/qr.json"){
    url <- getUrl()
    qr <- getQRCode(url)
    qr_json <- jsonlite::toJSON(list(qr=qr, url=url), auto_unbox=TRUE)
    return(list(status = 200L,
                headers = list('Content-Type' = 'application/json'),
                body = qr_json ))
  }else if(path == "/data.json"){
    if(is.null(pkg.env$vr_data_json))
      pkg.env$vr_data_json <- defaultDataJson()
    return(list(status = 200L,
                headers = list('Content-Type' = 'application/json'),
                body = pkg.env$vr_data_json ))
  }else{
    real_filename <- system.file(path,package="plotVR")
    real_filename <- sub("../","", real_filename,fixed=T)
    log.info(" changed to ",real_filename)

    if(!file.exists(real_filename)){
      log.info("Does not exist!!")
      return(list(status=404L,headers=list('Content-Type'="text/plain"),body="Nothing here."))
    }
    content_type <- 'text/html'
    if(tools::file_ext(path)=="png") content_type <- "image/png"
    if(tools::file_ext(path)=="js") content_type <- "application/javascript"
    return(list(status = 200L,
                headers = list('Content-Type' = content_type),
                body = list(file=real_filename)))
  }
}

handle_upload_data <- function(req){
  log.info('Got POST Content-type;', req$CONTENT_TYPE)
  input_str <- req$rook.input$read_lines()
  json_data <- paste(input_str, collapse="\n")
  pkg.env$vr_data_json <- json_data
  log.info('Having new data from POST -- nchar:',paste0(nchar(json_data),collapse=','),'data:',json_data)
  #ws()log.info(pkg.env$vr_data_json,fill=T)

  broadcastRefresh()

  log.info('got data and broadcasted refresh')
  return(list(status = 200L,
              headers = list('Content-Type' = 'text/plain'),
              body = "OK"))
}

handle_echo <- function(wsUrl){
  list(status = 200L,
                headers = list('Content-Type' = 'text/html'),
                body = paste(
                  sep = "\r\n",
                  "<!DOCTYPE html>",
                  "<html>",
                  "<head>",
                  '<style type="text/css">',
                  'body { font-family: Helvetica; }',
                  'pre { margin: 0 }',
                  '</style>',
                  "<script>",
                  sprintf("var ws = new WebSocket(%s);", wsUrl),
                  "ws.onmessage = function(msg) {",
                  '  var msgDiv = document.createElement("pre");',
                  '  msgDiv.innerHTML = msg.data.replace(/&/g, "&amp;").replace(/\\</g, "&lt;");',
                  '  document.getElementById("output").appendChild(msgDiv);',
                  "}",
                  "function sendInput() {",
                  "  var input = document.getElementById('input');",
                  "  ws.send(input.value);",
                  "  input.value = '';",
                  "}",
                  "</script>",
                  "</head>",
                  "<body>",
                  '<h3>Send Message</h3>',
                  '<form action="" onsubmit="sendInput(); return false">',
                  '<input type="text" id="input"/>',
                  '<h3>Received</h3>',
                  '<div id="output"/>',
                  '</form>',
                  "</body>",
                  "</html>"
                )
    )
}


processWS <- function(thesock) {
  log.info("got WebSocket")
  self <- as.environment(list(ws=thesock, isDevice=F, isController=F, hasFocus=F))
  pkg.env$clients <- c(pkg.env$clients, list(self))
  thesock$onMessage(function(binary, message) onMessage(binary, message, self))
  thesock$onClose(function() onClose(self) )
}

onMessage <- function(binary, message, self) {
    log.info("got WS message",message)
    body <- jsonlite::fromJSON(message)
    keys <- ls(body)
    if('shutdown' %in% keys){
      # FIXME broadcast it?
      stopBlockingServer()
      return()
    }
    sendStatus <- F
    if('focus' %in% keys){
      self$hasFocus <- body$focus
      sendStatus <- T
    }
    if('controller' %in% keys){
      self$isController <- body$controller
      sendStatus <- T
    }
    if('device' %in% keys){
      self$isDevice <- body$device
      sendStatus <- T
    }
    if(sendStatus){
      broadcastStatus()
    }else{
      broadcast(message)
    }
  }

onClose <- function(self){
  handle <- self$thesock$.handle
  log.info("Closing websocket: ",handle)
  # str(thesock)
  # remove that client from all_clients
  handles <- sapply(pkg.env$clients, function(x) x$thesock$.handle)
  pkg.env$all_clients <- pkg.env$all_clients[handles != handle]
  log.info("Closed websocket: ",handle)
}

app <- list(call=process_request, onWSOpen = processWS)

# ' Start the WebServer in a blocking version.
startBlockingServer <- function(host="0.0.0.0", port=2908, interruptIntervalMs=ifelse(interactive(), 100, 1000)){

  log.info("Welcome to the PlotVR Server")
  server <- httpuv::startServer(host, port, app)
  pkg.env$server_handle <- server

  on.exit(stopBlockingServer())
  pkg.env$stopped <- FALSE
  pkg.env$interruptIntervalMs <- interruptIntervalMs

  cat('Now do: browseURL("http://',host,':',port,'/")\n')

  while (!pkg.env$stopped) {
    httpuv::service(interruptIntervalMs)
    Sys.sleep(0.001)
  }

  log.info("Stopped blocking server")
}


stopBlockingServer <- function(server=pkg.env$server_handle){
  if(is.null(server))
    return()
  log.info("Stopping Blocking Server ",pkg.env$server_handle,"... ")

  if(server == pkg.env$server_handle){
    pkg.env$server_handle <- NULL

    # set stopped flag and give time
    pkg.env$stopped <- TRUE
    Sys.sleep(2*pkg.env$interruptIntervalMs/1000)
    log.info("waited", 2*pkg.env$interruptIntervalMs/1000, 's')
  }

  try(httpuv::stopServer(server))

  log.info("succeeded\n")
}

broadcastRefresh <- function(data){
  if(!missing(data)){
    pkg.env$vr_data_json <- data
  }
  broadcast(list(key='reload_data'))
}

broadcastStatus <- function(data){
  print(length(pkg.env$clients))
  status <- list(
    numDevices=sum(sapply(pkg.env$clients, `[[`, 'isDevice')),
    numControllers=sum(sapply(pkg.env$clients, `[[`, 'isController')),
    numFocus=sum(sapply(pkg.env$clients, `[[`, 'hasFocus'))
  )
  broadcast(list(status=status))
}

broadcast <- function(message){
  if(!is.character(message))
    message <- jsonlite::toJSON(message, auto_unbox=TRUE, pretty=TRUE)
  for(client in pkg.env$clients){
    log.info("Sending", message, "to") #,as.character(ws))
    try(client$ws$send(message))
  }
}
